### PR TITLE
Use Kafka change feed for blob deletion pillow

### PR DESCRIPTION
**Pull request by millerdev**
A single pillow now processes both application and main db blob deletions.

Follow up on https://github.com/dimagi/commcare-hq/commit/ceaedba69cf70e21dcc8bc8c6050714ba3a56440#commitcomment-17248339

@czue does this look like it will work?

### CODE REVIEW

File: corehq/blobs/pillow.py [hunk 1]
```diff
@@ -19,46 +14,28 @@
...
-    def __init__(self, blob_db, bucket_base):
+    def __init__(self, blob_db):
         super(BlobDeletionProcessor, self).__init__()
         self.blob_db = blob_db
-        self.bucket_base = bucket_base
 
     def process_change(self, pillow_instance, change):
         if change.deleted:
-            bucket = join(self.bucket_base, change.id)
+            bucket = "{}/{}".format(change.meta.data_source_name, change.id)
```
Comment (millerdev) [inline corehq/blobs/pillow.py]: @czue: I couldn't find any other references to `change.meta.data_source_name` in cchq code, so I'm trusting you on this one.

Comment (czue) [inline corehq/blobs/pillow.py]: sorry I think it should be `change.metadata`.

here's [where it get's set](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/change_feed/consumer/feed.py#L146)
and I think the document store definitely has the right DB name or all sorts of stuff wouldn't work

Comment (czue) [inline corehq/blobs/pillow.py]: and sorry, here's [where the kafka message gets populated from couch](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/change_feed/pillow.py#L33).

though looking at this I think there might be a problem with my suggestion which is that I don't think hard deletions are properly pushed to kafka. I think to fix that we need to check for `change.deleted` and make a metadata object with the right id and data source info so that the other pillows can pick it up. let me know if that makes sense and is something you'd be willing to add to this PR - otherwise I think we can put it on the backlog (or maybe @snopoke will volunteer to give it a shot)

File: corehq/blobs/pillow.py [hunk 2]
```diff
@@ -19,46 +14,28 @@
...
-    if change_feed is None:
-        change_feed = CouchChangeFeed(couch_db, include_docs=False)
+    checkpoint = PillowCheckpoint(pillow_id)
     return ConstructedPillow(
         name=pillow_id,
         checkpoint=checkpoint,
-        change_feed=change_feed,
-        processor=BlobDeletionProcessor(get_blob_db(), couch_db.dbname),
+        change_feed=KafkaChangeFeed(
+            topics=[topics.META, topics.APP],
```
Comment (czue) [inline corehq/blobs/pillow.py]: i think once form attachments go in the blobdb we would need to add `topics.FORM` and `topics.FORM_SQL` to delete form attachments. but thinking about that we might not want to since those are primary data. we may need some configuration that controls whether to actually delete blobs based on something like the doc type.

**Comment by czue** [comment]
also seems like it'd be good to have tests for this.

**Comment by czue** [comment]
or maybe there are tests? build is failing:
```
======================================================================
ERROR: test_process_change_with_deleted_document (corehq.blobs.tests.test_pillow.TestBlobDeletionProcessor)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/blobs/tests/test_pillow.py", line 13, in setUp
    self.processor = BlobDeletionProcessor(self.db, db_name)
TypeError: __init__() takes exactly 2 arguments (3 given)
-------------------- >> begin captured logging << --------------------
```
also need to update the prod pillow list for tests:
```
======================================================================
FAIL: test_instantiate_all (testapps.test_pillowtop.tests.test_settings.PillowtopSettingsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_settings.py", line 34, in test_instantiate_all
    self.assertEqual(len(all_pillow_configs), len(expected_meta))
AssertionError: 46 != 47
----------------------------------------------------------------------
```

File: corehq/blobs/pillow.py [hunk 3]
```diff
@@ -19,46 +14,28 @@
...
+            group_id='blob-deletion-group',
+        ),
+        processor=BlobDeletionProcessor(get_blob_db()),
         change_processed_event_handler=PillowCheckpointEventHandler(
```
Comment (snopoke) [inline corehq/blobs/pillow.py]: This should be a `MultiTopicCheckpointEventHandler`

File: corehq/blobs/pillow.py (hunk 1 above)
Comment (millerdev) [inline corehq/blobs/pillow.py]: If I'm understanding this correctly [soft and hard deletions are always sent to kafka as soft deletions](https://github.com/dimagi/commcare-hq/blob/58e88da8458f1a271849bd02e2beb2b153ed70a2/corehq/apps/change_feed/pillow.py#L47). I'm not even sure what would need to be done to fix that. Would you propose to a `deleted` field to [`ChangeMeta`](https://github.com/dimagi/commcare-hq/blob/58e88da8458f1a271849bd02e2beb2b153ed70a2/corehq/ex-submodules/pillowtop/feed/interface.py#L7-L20) so kafka can distinguish between hard and soft deletions?

File: corehq/blobs/pillow.py (hunk 2 above)
Comment (millerdev) [inline corehq/blobs/pillow.py]: Are forms ever hard-deleted? If not, I think `BlobDeletionProcessor` will not delete the blobs.

File: corehq/blobs/pillow.py (hunk 3 above)
Comment (millerdev) [inline corehq/blobs/pillow.py]: 58e88da

File: corehq/blobs/pillow.py (hunk 1 above)
Comment (czue) [inline corehq/blobs/pillow.py]: this is a bit of a mess (my fault). my intention was that by the time they get to the pillow soft and hard deletions could be treated the same, so the kafka feed populates `deleted` with the value from the `ChangeMeta.is_deleted` [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/change_feed/consumer/feed.py#L160).

I think the long term thing to do would be to get rid of one of them (and also maybe add a way to flag soft deletions if we do need to distinguish between them as you suggest). But for now I think this should work as long as we update the kafka feed to publish hard deletions and don't process deletions from forms.

File: corehq/blobs/pillow.py (hunk 2 above)
Comment (czue) [inline corehq/blobs/pillow.py]: Users can't hard delete forms through the UI so it's only scripts or things like that that could delete them. However I'd be worried about soft deletions (I think a good reason to add that to the `meta` object)

File: corehq/blobs/pillow.py (hunk 1 above)
Comment (millerdev) [inline corehq/blobs/pillow.py]: > But for now I think this should work as long as we update the kafka feed to publish hard deletions

Do you mean "update the kafka feed to *only* publish hard deletions" as in *do not push soft deletions*?

> and don't process deletions from forms.

It's a unclear why we wouldn't want to delete forms if the corresponding data in couch/postgres was being hard-deleted. I'm not sure how we'd ever use blobs that were not referenced by a record in another database.

Comment (czue) [inline corehq/blobs/pillow.py]: sorry I think i am creating confusion here and possibly github is not the best place for this conversation. here's my quick summary of the situation.

I think the ideal future state is:

- pillows publish both soft and hard deletions and differentiate between them
- blobs are deleted on all hard deletions
- blobs are not deleted on all soft deletions (including forms)

I think the current state is:

- pillows inconsistently publish hard deletions (couch pillows do, kafka pillows mostly don't)
- hard and soft deletions are not differentiable
- blobs are deleted on hard deletions in couch pillows

I think this PR changes the last thing to:

- blobs are deleted on hard and soft deletions that happen to get published to the `META` and `APP` topics (which may not be all of them because of the first bullet in "current state".

which is probably a net worse change. So I'm tempted to propose as next steps:

- we put in the backlog to fix the hard/soft deletion publishing
- when that's done we come back to this and it should be ready to drop in.

thoughts?

Comment (millerdev) [inline corehq/blobs/pillow.py]: > I think this PR changes the last thing to:
> * blobs are deleted on hard and soft deletions...
>
> which is probably a net worse change.

Completely agree with this, which is why I added the *hold off on merge* tag (was not planning to remove it until that was resolved).

I think your proposed solutions sound good. +1

Comment (czue) [inline corehq/blobs/pillow.py]: FYI, I looked into this a little bit today and I think it's more complicated than this. I don't think we have any clear way to deal with hard deletions in SQL, and even in couch it doesn't look like it obviously works (I don't think that hard deletions have enough information to know the doc type which means we don't know how to publish it to kafka). For the blobdb this would be fine because you only use the ID, but for other pillows it makes things complicated. I think we might need to add a "deleted" topic to kafka and then publish all hard deletions to that.

Anyway, bottom line is that I think there's a bit of design work required to sort this out and not sure how much of a priority it is so guessing this might linger in limbo for a while.

Comment (millerdev) [inline corehq/blobs/pillow.py]: Cool, thanks for looking into it.

There is another proposal I intend to follow up on at some point that would render this irrelevant: make the blob db append-only (never delete blobs). The hope is to make it unnecessary to backup the riak cluster, since it already maintains redundancy and therefore doesn't require backups as long as nothing is ever deleted. I'd like to do a few calculations to do a rough comparison of the (cost of retaining) space for deleted blobs against the cost of maintaining backups for the riak cluster.

**Comment by millerdev** [comment]
re-open if/when this becomes easier/possible.
